Directly return from final sha3/keccak_final if no bytes are requested

p-steuer opened the pull request:
Reference implementation in keccak1600.c does not touch output buffer
if output lenght is zero while some asm implementations do.

A regression test for this issue is added to evp_test.c

Fixes https://github.com/openssl/openssl/issues/9431

I need help to verify the fix for architectures i have no access.

Already verified for: s390.

- [ ] documentation is added or updated
- [x] tests are added or updated

kroeckx commented:
Let me know if I should this this on some platforms. I have access to all Debian architectures.

kroeckx commented:
Let me know if I should this this on some platforms. I have access to all Debian architectures.

t8m commented:
I suppose all the assembler code implementations have this deficiency.

p-steuer commented:
no fix needed, verified:
- keccak1600-x86_64.pl

fixed, verified:
- keccak1600-s390x.pl
- keccak1600-ppc64.pl
- keccak1600p8-pcc.pl
- keccak1600-x86_64.pl
- keccak1600-avx2.pl

no fix needed, not verified:
- keccak1600-mmx.pl

fixed, not verified:
- keccak1600-avx512.pl
- keccak1600-avx512vl.pl

todo:
- keccak1600-armv4.pl
- keccak1600-armv8.pl
- keccak1600-c64x.pl

Please hep with the still to be verified platforms, if you have access or an emulation set up. 

I dont feel comfortable making changes to the arm and c64 asm. @dot-asm could you help here ?

kroeckx commented:
Isn't this easier to fix in the C code, for all platforms at the same time?

p-steuer commented:
certainly easier but is it the right thing to do ? i think the other asm modules handle len=0 input. opinions ?

kroeckx commented:
I would do the check as early as possible.

dot-asm commented:
Here is "philosophical" question. Where would you stop? I mean if you test for buffer length, then why not test for pointers' validity[!] as well? And block size? What I'm trying to say is that there **is** a reasonable limit for amount of checks, and it is **not** inappropriate to make it caller's responsibility. Observation is that some assembly subroutines behave differently from reference implementation. So what? One can legitimately claim that despite that reference SHA3_squeeze implementation does nothing if zero length is passed, doing so constitutes undefined behaviour. It won't be a contradiction.

Also, keep in mind that this is private interface, application code won't be making the direct call. So question is where does length argument come from? It's either constant or value set with EVP_MD_CTRL_XOF_LEN. But then wouldn't it be more appropriate to vet the value in shake_ctrl? Because that's where you can communicate to application whether or not it's suitable? As opposite to choosing to opt for something application has no way of figuring out as effectively suggested here?

Just in case, as for vetting value in shake_ctrl. Note that it's `int` that is assigned to `size_t`. So that non-zero check wouldn't be right, one has to check for positivity.

But in either case, let's say that consensus would still be to check for non-zero length in all SHA3_squeeze. Since we don't check for pointers' validity, would it be appropriate to assume that pointer to A[][] is always valid? Why question? Well, it's only appropriate to aim for minimizing amount of branches. This in turn means that if a branch has to be added, it would be appropriate to position it some place where it would be less likely to be evaluated. In this case to squeeze_tail, which is actually an [highly] unlike path. This in turn would mean that for example on PPC A[][] would be always dereferenced... What I'm getting at is that it's possible to avoid additional misprediction penalty for most common scenario, which should be viewed as desirable goal.

p-steuer commented:
after comments from @kroeckx and @dot-asm , i reworked this one:
- passing zero length to SHA3_squeeze is now undefined behavior
- the zero length check is moved to sha3/keccak_final

p-steuer commented:
ping. after the change i think i can remove the test code: we dont test for out-of-bounds writes in other places and this can be left to address sanitizers ..

levitte commented on test/evp_test.c:
```diff
@@ -404,6 +404,28 @@ static int digest_test_run(EVP_TEST *t)
 
     if (EVP_MD_flags(expected->digest) & EVP_MD_FLAG_XOF) {
+        EVP_MD_CTX *mctx_cpy;
+        char dont[] = "touch";
```
This made me giggle :-)

p-steuer commented:
pushed to master: a890ef833d114da3430c2f2efd95e01714704d34 3ce46435e6ebed69bec0fa3454cc195ced426d42
cherry-picked to 1.1.1: 6087d4a6fb1877c668dcf934da6a2f9e402dff1d efc62e661714a4bba3040a12da4e4a0df8855283
